### DISCUSSION TITLE
How to choose GPU?

### DISCUSSION THREAD

**Discussion by elchupacabrinski** [vladmandic/sdnext#977] [Q&A]
I have 2 GPUs in my system and I'd like to run 2 installs simultaneously.

However when I add  "set CUDA_VISIBLE_DEVICES=1" to webui-user.sh I get the following error when trying to launch the UI:
: error: unrecognized arguments: CUDA_VISIBLE_DEVICES=1

Can anyone help me out here?

**Comment by Tillerz** [comment]
To force it to use cuda (nvidia), you use --use-cuda ... or for amd gpu you use --use-rocm. 

I don't think you can make it use only gpu a or gpu b when it's the same gpu type.

Looks like this without any manually added parameters:

```21:07:32,539 | DEBUG | X:\automatic\installer.py | Torch overrides: cuda=False rocm=False ipex=False diml=False
21:07:32,539 | DEBUG | X:\automatic\installer.py | Torch allowed: cuda=True rocm=True ipex=True diml=True
21:07:32,542 | INFO  | X:\automatic\installer.py | nVidia CUDA toolkit detected
21:07:32,641 | DEBUG | X:\automatic\installer.py | Package version found: torch 2.0.1+cu118
21:07:32,641 | DEBUG | X:\automatic\installer.py | Package version found: torchvision 0.15.2+cu118
21:07:33,595 | INFO  | X:\automatic\installer.py | Torch 2.0.1+cu118
21:07:33,604 | INFO  | X:\automatic\installer.py | Torch backend: nVidia CUDA 11.8 cuDNN 8700
21:07:33,606 | INFO  | X:\automatic\installer.py | Torch detected GPU: NVIDIA GeForce RTX 4090 VRAM 24564 Arch (8, 9) Cores 128
21:07:33,608 | DEBUG | X:\automatic\installer.py | Package version found: tensorflow 2.12.0```

So it finds the stuff and uses it.

**Reply by elchupacabrinski** [reply]
Thank you, but afaik it does only use one GPU by default doesn't it?
I've read multiple times that people configured Auto1111 to use a specific GPU so they can run 2 GUIs at once but somehow I can't seem to get it to work

**Comment by vladmandic** [comment]
> Thank you, but afaik it does only use one GPU by default doesn't it?

By default it uses all GPUs, but there is no way to split single task accross multiple GPUs. if you run with high batch-size values, it should be split since each is then parallel. That's the theory as I don't have multiple GPUs.

> I've read multiple times that people configured Auto1111 to use a specific GPU so they can run 2 GUIs at once but somehow I can't seem to get it to work

Yes, environment variable `CUDA_VISIBLE_DEVICES` applies here as well.
It seems you have a syntax error something, its not that variable itself cannot be used.
i don't know what's your OS or shell used, so can't say much more.

**Reply by elchupacabrinski** [reply]
Thanks for clearing that up. If it does use both by default I don't need to change anything. Nice :)

**Reply by vladmandic** [reply]
yes, but only for workloads that can be parallelized - for single image generation, it just cant.

**Comment by Tillerz** [comment]
Oh wait, I see .sh up there, are you on linux? The interwebs says:

```
export COMMANDLINE_ARGS="--device-id=1"
```

for linux for the second GPU. --device-id=0 is for the first GPU, it seems.

**Reply by vladmandic** [reply]
you can do simply
`export CUDA_VISIBLE_DEVICES=1 ./webui.sh`

**Comment by scooller** [comment]
I have a Laptop with a integrated card Nvidia GTX 970M and intel 4600 hd graphics...
and I can't get it to select the Nvidia card, CPU is always selected

I add:
set COMMANDLINE_ARGS=--medvram --device-id 0
set CUDA_VISIBLE_DEVICES=0,1

call webui.bat

I try --device-id 1, --device-id 0, --device-id=1, --device-id=0, etc... nothing works for me

**Comment by jasonmhead** [comment]
For selecting a second GPU in Windows, what would be the best way?
I tried --device-id 1 but it still used my first GPU

What file and modifications should be done to select a second GPU?
